python/TestHarness/schedulers/Scheduler.py

```diff
@@ -339,60 +350,59 @@ def schedule(self, testers):
             non_runnable_jobs.update(additionally_skipped)
             job_dag.delete_node_if_exists(job)
 
-        # Get a count of all the items still in the DAG. These will be the jobs that ultimately are queued
-        runnable_jobs = job_dag.size()
-
         # Make sure we didn't drop a tester somehow
-        if len(non_runnable_jobs) + runnable_jobs != len(testers):
+        if len(non_runnable_jobs) + job_dag.size() != len(testers):
             raise SchedulerError('Runnable tests in addition to Skipped tests does not match total scheduled test count!')
 
         # Inform derived schedulers of the jobs we are skipping immediately
         self.reportSkipped(non_runnable_jobs)
 
-        # Assign a status thread to begin work on any skipped/failed jobs
-        self.queueJobs(status_jobs=non_runnable_jobs)
-
         # Allow derived schedulers to modify the dag before we launch
         # TODO: We don't like this, and this will change when we move to better DAG handling.
-        if runnable_jobs:
+        if job_dag.size():
             self.preLaunch(job_dag)
 
         # Build our list of runnable jobs and set the tester's status to queued
-        job_list = []
-        if runnable_jobs:
-            job_list = job_dag.ind_nodes()
-            for job_container in job_list:
-                tester = job_container.getTester()
+        if job_dag.size():
+            runnable_jobs = job_dag.ind_nodes()
+            for job in runnable_jobs:
+                tester = job.getTester()
                 tester.setStatus('QUEUED', tester.bucket_pending)
 
-        # Queue runnable jobs
-        self.queueJobs(run_jobs=job_list)
+        # Queue generated jobs
+        self.queueJobs(run_jobs=runnable_jobs, status_jobs=non_runnable_jobs)
 
     def waitFinish(self):
         """
         Block while the job queue is not empty. Once empty, this method will begin closing down
         the thread pools and perform a join. Once the last thread exits, we return from this
         method.
 
-        There are two thread pools in play; the Tester pool which is performing all the tests,
+        There are two thread pools in play; the Run pool which is executing all the testers,
         and the Status pool which is handling the printing of tester statuses. Because the
         Status pool will always have the last item needing to be 'printed', we close and join
-        the Tester pool first, and then we do the same to the Status pool.
+        the Run pool first, and then, we close the Status pool.
         """
-        while self.job_queue_count > 0:
-            # One of our children died :( so exit uncleanly
-            if self.error_state:
-                self.killRemaining()
-                return
-            sleep(0.5)
+        try:
+            while self.job_queue_count > 0:
+                # One of our children died :( so exit uncleanly (do not join)
+                if self.error_state:
+                    return
```

Comment:
What happens to currently running threads? Should this do a `killRemaining()` before returning?